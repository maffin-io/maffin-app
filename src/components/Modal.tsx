import React from 'react';

export type ModalProps = {
  open: boolean;
  setOpen: Function;
  title?: string;
  children: React.ReactNode;
};

export default function Modal({
  open, setOpen, title, children,
}: ModalProps): JSX.Element {
  if (!open) {
    return <span />;
  }

  return (
    <div
      className="fixed inset-0 bg-white bg-opacity-30 overflow-y-auto h-full w-full z-50"
      id="my-modal"
    >
      <div
        className="relative top-20 mx-auto p-5 w-1/3 shadow-lg rounded-md bg-gunmetal-700"
      >
        <button
          type="button"
          className="float-right"
          onClick={() => setOpen(false)}
        >
          X
        </button>
        {
          title
          && (
            <div className="w-full pb-4 border-b border-gunmetal-600">
              <span className="text-lg">
                {title}
              </span>
            </div>
          )
        }
        {children}
      </div>
    </div>
  );
}
