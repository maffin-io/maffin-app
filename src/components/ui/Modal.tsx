import React from 'react';
import RModal from 'react-modal';

export type ModalProps = {
  className?: string;
  triggerContent: string | JSX.Element;
  triggerClassName?: string;
  showClose?: boolean;
} & React.PropsWithChildren;

export interface ModalRef {
  openModal: () => void;
  closeModal: () => void;
}

function Modal({
  className = 'modal',
  triggerClassName = '',
  triggerContent,
  showClose = false,
  children,
}: ModalProps, ref: React.Ref<ModalRef>): JSX.Element {
  const [isOpen, setIsOpen] = React.useState<boolean>(false);

  // Expose actions to the parent
  React.useImperativeHandle(ref, () => ({
    openModal: () => setIsOpen(true),
    closeModal: () => setIsOpen(false),
  }));

  return (
    <>
      <RModal
        isOpen={isOpen}
        overlayClassName="overlay"
        className={className}
        onRequestClose={() => setIsOpen(false)}
      >
        {
          showClose
          && (
            <button
              type="button"
              className="float-right"
              onClick={() => setIsOpen(false)}
            >
              X
            </button>
          )
        }
        {children}
      </RModal>

      <button
        type="button"
        onClick={() => setIsOpen(!isOpen)}
        className={triggerClassName}
      >
        {triggerContent}
      </button>
    </>
  );
}

export default React.forwardRef(Modal);
