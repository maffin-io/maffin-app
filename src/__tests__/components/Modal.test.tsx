import React from 'react';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import Modal from '@/components/Modal';

describe('Modal', () => {
  it('returns span if not open', () => {
    const { container } = render(
      <Modal
        open={false}
        setOpen={() => {}}
      >
        <div />
      </Modal>,
    );

    expect(container.innerHTML).toEqual('<span></span>');
  });

  it('renders as expected', () => {
    const { container } = render(
      <Modal
        title="title"
        open
        setOpen={() => {}}
      >
        <div />
      </Modal>,
    );

    expect(container).toMatchSnapshot();
  });

  it('calls setOpen on close', async () => {
    const user = userEvent.setup();
    const mockSetOpen = jest.fn();
    render(
      <Modal
        title="title"
        open
        setOpen={mockSetOpen}
      >
        <div />
      </Modal>,
    );

    screen.getByText('title');
    await user.click(screen.getByText('X'));

    expect(mockSetOpen).toBeCalledTimes(1);
    expect(mockSetOpen).toBeCalledWith(false);
  });
});
