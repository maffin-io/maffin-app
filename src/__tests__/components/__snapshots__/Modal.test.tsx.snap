// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal renders as expected 1`] = `
<div>
  <div
    class="fixed inset-0 bg-white bg-opacity-30 overflow-y-auto h-full w-full z-50"
    id="my-modal"
  >
    <div
      class="relative top-20 mx-auto p-5 w-1/3 shadow-lg rounded-md bg-gunmetal-700"
    >
      <button
        class="float-right"
        type="button"
      >
        X
      </button>
      <div
        class="w-full pb-4 border-b border-gunmetal-600"
      >
        <span
          class="text-lg"
        >
          title
        </span>
      </div>
      <div />
    </div>
  </div>
</div>
`;
